lib/src/component/with_transition.dart

```diff
@@ -0,0 +1,426 @@
+  ///
+  /// A transition may not always occur when the state moves from SHOWING to HIDING
+  /// if the PRE_SHOWING-->SHOWING-->HIDING transition happens back-to-back.
+  ///
+  /// Better to not always transition when the user is ninja-toggling a transitionable
+  /// component than to break state changes waiting for a transition that will never happen.
+  var _transitionNotGuaranteed = false;
+
+  /// Timer used to determine if a transition timeout has occurred.
+  Timer _transitionEndTimer;
+
+  /// Stream for listening to `transitionend` events on the node.
+  StreamSubscription _endTransitionSubscription;
+
+  final _transitionNodeRef = createRef<Element>();
+
+  @override
+  get defaultProps => (newProps()
+    ..addProps(v2.TransitionPropsMixin.defaultProps)
+    ..isShown = false
+    ..transitionTimeout = const Duration(seconds: 1)
+    ..childPropsByPhase = const {
+      v2.TransitionPhase.PRE_SHOWING: {},
+      v2.TransitionPhase.SHOWING: {},
+      v2.TransitionPhase.SHOWN: {},
+      v2.TransitionPhase.HIDING: {},
+      v2.TransitionPhase.HIDDEN: {},
+    }
+  );
+
+  @override
+  get initialState => (newState()
+    ..$transitionPhase = _transitionPhaseDerivedFromProps(props)
+  );
+
+  @override
+  getDerivedStateFromProps(Map nextProps, Map prevState) {
+    final tNextProps = typedPropsFactory(nextProps);
+    final tPrevState = typedStateFactory(prevState);
+    if (tNextProps.isShown && _isOrWillBeHidden(tPrevState.$transitionPhase)) {
+      return _stateToBeginShowing(tNextProps, tPrevState.$transitionPhase);
+    } else if (!tNextProps.isShown && _isOrWillBeShown(tPrevState.$transitionPhase)) {
+      return _stateToBeginHiding(tNextProps, tPrevState.$transitionPhase);
+    }
+
+    return null;
+  }
+
+  @override
+  void componentDidUpdate(Map prevProps, Map prevState, [dynamic snapshot]) {
+    if (state.$transitionPhase == typedStateFactory(prevState).$transitionPhase) return;
+
+    _transitionNotGuaranteed = false;
+
+    if (state.$transitionPhase != TransitionPhase.SHOWING) {
+      // Allows the `WithTransitionComponent` to handle state changes that interrupt state
+      // changes waiting on `transitionend` events.
+      _cancelTransitionEventListener();
+    }
+
+    switch (state.$transitionPhase) {
+      case TransitionPhase.PRE_SHOWING:
+        _handlePreShowing();
+        break;
+      case TransitionPhase.SHOWING:
+        // NOTE: props.onWillShow is called within _handleShow in the handler that
+        // triggers an update to `$transitionPhase` that will result in this effect being processed.
+        break;
+      case TransitionPhase.HIDING:
+        // NOTE: props.onWillHide is called within _handleHide in the handler that
+        // triggers an update to `$transitionPhase` that will result in this effect being processed.
+        _transitionNotGuaranteed = state.$transitionPhase == TransitionPhase.SHOWING;
```

Comment:
If I change it to `tPrevState` - two tests fail.

Reply:
So perhaps this flag isn't needed anymore? If the code isn't changed to prevState, then the flag should be eliminated, because right now it's always `false`.

Do the AbstractTransition tests fail when that code I linked is changed to `_transitionNotGuaranteed = false;`?

Reply:
Yes

Reply:
Cool, sounds like this impl just doesn't need it, likely because of the different way state transitions work

Reply:
Removed in https://github.com/Workiva/over_react/pull/621/commits/e0208a6375576540b0fd0cccc5ffd2e5cc74f611